[[ UntangleVM ]] Work on restructuring core VM bytecode implementation.

runrevmark opened the pull request:
Work-in-progress on restructuring the bytecode loops in the VM.

Bytecodes are in individual classes (as a set of static methods) and break down their operations into more primitive ones executing on a Validate or Execute context.

runrevmark committed:
[[ UntangleVM ]] Work on restructuring core VM bytecode implementation.

WIP

peter-b commented:
@runrevmark It's great to have all the code in one place and hopefully it'll be much easier to write bytecode op implementations with the "do stuff, check at the end" model.

I was thinking that, for more complex bytecode implementations, it might be worth querying the context validity and possibly bailing out early, in order to simplify reasoning about the possibility that the context is invalid. I'm thinking specifically of the invoke implementation. WDYT?

runrevmark commented:
@peter-b: Is that necessary though? If an opcode bails, then it means that its a runtime (programmatic) error. The model I've gone for after discussions yesterday is that the context stores the error flag, all imperative actions will do nothing if it is set, all function actions (those which query state - e.g. FetchRegister) will return a sensible default in the case of an error (typically null). Indeed, if you look at the Invoke implementation, it does actually bail as soon as the context is in error. Basically all bytecodes consist of:

```
   prep for action
   do action
```

The 'prep for action' involves computations which will bail if an error value (e.g. null) is returned, and the 'do action' is usually the very last thing done on any code-path through the op.

peter-b commented:
@runrevmark No, I don't think it's necessary.  Carry on! 👍

runrevmark committed:
WIP

runrevmark committed:
WIP pre combine context and op classes

runrevmark committed:
WIP

runrevmark committed:
WIP - renamed .cpp.h -> .hpp, moved to use 'Visitor' / 'Visit' rather than Context / operator ()

peter-b commented on libscript/src/script-bytecode.hpp:
```diff
@@ -0,0 +1,871 @@
+		else
+		{
+			ctxt.ReportInvalidArity();
+		}
```
This could be constructed as `switch (ctxt.GetArity())`.

peter-b commented on libscript/src/script-bytecode.hpp:
```diff
@@ -0,0 +1,871 @@
+		else
+		{
+			__MCScriptUnreachable__("invalid definition kind in internal handler-ref");
+		}
```
This looks like a `switch (t_context->definition->kind)`

runrevmark committed:
WIP - removed old BytecodeInfo table - replaced with per Op struct Describe method

peter-b commented on libscript/src/script-validate.hpp:
```diff
@@ -0,0 +1,299 @@
+	if (p_register >= m_register_limit)
+	{
+		m_register_limit = p_register + 1;
+	}
```
Is there a hard limit on register count? Can I construct a bytecode block that makes `m_register_limit` overflow (and, if I can, is that a problem)?

runrevmark replied:
Bytecode blocks only allow variable definitions which are always allocated a register in order from the sequence of definitions. The actual number of registers needed is computed as the maximum of all register indicies in the bytecode of the handler... This means it would be possible to overflow that - the maximum register index needs to be UINDEX_MAX-1. I'm thinking of making it so that the compiler emits the number of registers needed for a handler, and then the validation check is that there are no indicies greater than that and _all_ indicies are referenced at least once in the bytecode. This concern goes away completely when we move to a stack-based format though, so I'll just make CheckRegister check that p_register < UINDEX_MAX for now.

peter-b replied:
Just adding the check sounds good to me.

runrevmark committed:
WIP

Added CheckMinimum/MaximumArity to validate context.

Replaced switch-like if/elseifs with switch.

peter-b commented on libscript/src/script-validate.hpp:
```diff
@@ -0,0 +1,299 @@
+							MCScriptHandlerDefinition *definition,
+							uindex_t address,
+							uindex_t *arguments,
+							uindex_t argument_count);
```
Probably being daft, but I can't seem to find the implementation of this constructor anywhere in this patch.

runrevmark replied:
That's because I've not finished that bit yet ;)

peter-b commented on libscript/src/script-bytecode.hpp:
```diff
@@ -0,0 +1,904 @@
+		// input type for an argument is equal to expected type, then this is a
+		// score of zero. If input type for an argument conforms to expected type, then
+		// this is a score of 1. The method with the lowest score is chosen, with
+		// priority given to the first in the list in the case of a tie.
```
I still feel like a score tie should be an error, because it makes runtime behaviour sensitive to the order in which the syntax modules happened to be compiled.

runrevmark committed:
WIP

Make CheckRegister ensure index < UINDEX_MAX.

runrevmark committed:
WIP

Restructured templates for ForEach / Dispatch to take Function + Arg.

Various fixes to get compilation working and to pass LCB test suite.

(Validation and generation has been refactored - executino is still
to be finished).

runrevmark committed:
WIP

runrevmark committed:
WIP

runrevmark committed:
WIP

runrevmark committed:
WIP

runrevmark committed:
WIP

runrevmark committed:
WIP

runrevmark committed:
WIP - tweaks to libfoundation MCError constructors

runrevmark committed:
WIP

runrevmark committed:
WIP

runrevmark committed:
WIP

runrevmark committed:
WIP

runrevmark committed:
WIP

runrevmark commented:
Closing as I've repushed a new PR with a cleaner commit history.
